### ISSUE TITLE
Same national characters (Polish) are not displayed in large letters.

### ISSUE THREAD

**Issue by char11** [koreader/koreader#4958]
* KOReader version: v2019.04
* Device: Tolino Shine 2 HD

#### Issue
Same national characters (Polish) are not displayed in large letters.

#### Steps to reproduce
Download and open epub version https://www.dropbox.com/sh/x615847wi57xr8a/AABwuuag0YKCd7G-DcF_eiA0a?dl=0
Go to Table of contents and select "Pod skórą". 

![Reader](https://user-images.githubusercontent.com/25547161/56491056-88fcc280-64e7-11e9-9704-ee6ca247bc7e.png)

All characters should be large (tested for example in Calibre)
Other titles with Polish characters in this e-book also do not display correctly.

**Comment by Frenzie** [comment]
@poire-z It's a `text-transform: uppercase;`.

I think that's this function:

https://github.com/koreader/crengine/blob/7145b8625f2f24b49ee5391f578e2993c10f92fd/crengine/src/lvstring.cpp#L2675-L2704

Cf. https://github.com/koreader/crengine/pull/220

**Comment by char11** [comment]
I added this code to lStr_uppercase function but nothing happen.

    } else if ( ch==0x105 || ch==0x107 || ch==0x119 || ch==0x142 || 
                    ch==0x144 || ch==0x15b || ch==0x17a || ch==0x17c ) {
            str[i] = ch - 1;

**Comment by Frenzie** [comment]
`lStr_uppercase()` is the one you want for, well, uppercasing the entire string. ;-)

**Comment by poire-z** [comment]
That looks allright.
Guess you're in for debugging by adding some `printf ("yup\n")` to see if you go by there.

Well, this is some original crengine code (that I just picked the char codes from for the `lStr_capitalize` I added).
It's obviously missing many other char codes and rules, that the courageous one could pick from:
https://chromium.googlesource.com/native_client/nacl-newlib/+/489f1d8/newlib/libc/ctype/towupper.c
https://chromium.googlesource.com/native_client/nacl-newlib/+/489f1d8/newlib/libc/ctype/towlower.c

**Comment by Frenzie** [comment]
Is there a reason not to just use `towupper` directly instead of copying parts of it?

**Comment by poire-z** [comment]
Dunno. crengine mostly rewrote string functions, so aiming at being kind of standalone.
Dunno if the library providing it is available/usable (on all CR platforms, or only on ours?), if the types match or not... Could be tried.

**Comment by Frenzie** [comment]
It looks like a proper part of the standard to me:

https://en.cppreference.com/w/c/string/wide/towupper
https://en.cppreference.com/w/cpp/string/wide/towupper

Of course there might be implementation-specific variations depending on how well they implement the spec as well as C99 vs C11, but it seems plausible that our current modest implementation is likely to be almost invariably inferior to any built-in `towupper()` unless we were to copy the whole thing from GNU Libc.

(That being said, I wouldn't put it past Android's Bionic libc to be… lacking. Some small experiments might be warranted.)

**Comment by Frenzie** [comment]
Looks sufficiently sane to me since 2014: https://android.googlesource.com/platform/bionic/+/5a0aa3dee247a313f04252cf45608097695d5953

Of course we support 2011 Android 4. That looks something like this: https://android.googlesource.com/platform/bionic/+/29c7f0b4d18f812267c2194b85204e19e41d0387/libc/bionic/wchar.cpp

It should be fine; just worth double checking. :-)

**Comment by poire-z** [comment]
In the docs about these functions, they talk about locale having some effects, so I dunno if that would complicate things or not (last time I had to play with local was for sdcv on android, and it was not fun).

I prefer having the whole thing readable and deterministic in our code base thus :)

But feel free to have a go at it (I won't, not much time for crengining these days.)

**Comment by NiLuJe** [comment]
<wctype.h> is completely broken on Kobo (the FW doesn't ship the gcov tables), so anything even remotely related to widechar stuff is probably insidiously borked on Kobo.

**Comment by Frenzie** [comment]
Hm, fun. Well, I'd propose basically just lifting it from some source like GNU Libc then.

**Comment by NiLuJe** [comment]
c.f., https://github.com/NiLuJe/FBInk/blob/2bc5b2579ebf2ede4bcd93e8834e02ef909c2626/fbink_internal.h#L131 for reference and a rant about that ;p.

**Comment by NiLuJe** [comment]
The musl implementation might be an easier source (glibc code tends to rely on a large number of internal functions/macros).

**Comment by char11** [comment]
Diff for Polish characters. Thanks.
```
diff -ruN src/lvstring.cpp new/lvstring.cpp
--- src/lvstring.cpp	2019-04-21 13:35:24.037658935 +0200
+++ new/lvstring.cpp	2019-04-23 22:03:42.840096547 +0200
@@ -2621,6 +2621,9 @@
             str[i] = ch - 0x20;
         } else if ( ch>=0x3b0 && ch<=0x3cF ) {
             str[i] = ch - 0x20;
+        } else if ( ch==0x105 || ch==0x107 || ch==0x119 || ch==0x142 || 
+                    ch==0x144 || ch==0x15b || ch==0x17a || ch==0x17c ) {
+            str[i] = ch - 1;
         } else if ( (ch >> 8)==0x1F ) { // greek
             lChar16 n = ch & 255;
             if (n<0x70) {
@@ -2646,6 +2649,9 @@
             str[i] = ch + 0x20;
         } else if ( ch>=0x390 && ch<=0x3aF ) {
             str[i] = ch + 0x20;
+        } else if ( ch==0x104 || ch==0x106 || ch==0x118 || ch==0x141 || 
+                    ch==0x143 || ch==0x15a || ch==0x179 || ch==0x17b ) {
+            str[i] = ch + 1;
         } else if ( (ch >> 8)==0x1F ) { // greek
             lChar16 n = ch & 255;
             if (n<0x70) {
@@ -2687,6 +2693,9 @@
                 str[i] = ch - 0x20;
             } else if ( ch>=0x3b0 && ch<=0x3cF ) {
                 str[i] = ch - 0x20;
+            } else if ( ch==0x105 || ch==0x107 || ch==0x119 || ch==0x142 || 
+                        ch==0x144 || ch==0x15b || ch==0x17a || ch==0x17c ) {
+                str[i] = ch - 1;
             } else if ( (ch >> 8)==0x1F ) { // greek
                 lChar16 n = ch & 255;
                 if (n<0x70) {

```

**Comment by Frenzie** [comment]
@char11 Works well, thanks!

Sorry, I should've mentioned, but I grabbed this from @poire-z 's link the other day which also works:

```c
        else if ((ch >= 0x0101 && ch <= 0x012f) ||
          (ch >= 0x0133 && ch <= 0x0137) ||
          (ch >= 0x014b && ch <= 0x0177) ||
          (ch >= 0x01df && ch <= 0x01ef) ||
          (ch >= 0x01f9 && ch <= 0x021f) ||
          (ch >= 0x0223 && ch <= 0x0233) ||
          (ch >= 0x0247 && ch <= 0x024f))
        {
            if (ch & 0x01) {
                str[i] =  ch - 1;
            }
        }
```

But I don't currently have a clear vision going forward. So we could merge a small improvement while we figure it out. :-)

**Comment by char11** [comment]
It covers more languages - it's better. :-) Thanks again.

**Comment by Frenzie** [comment]
@char11 Bit early to close it. I couldn't find the issue anymore! ;-)

**Comment by Frenzie** [comment]
What @NiLuJe wrote is quite true btw. The musl code is very tidy, much easier to grok than glibc and up to date with Unicode 10 instead of 5.2 like @poire-z 's link. See http://git.musl-libc.org/cgit/musl/tree/src/ctype/towctrans.c

But instead of adapting from musl, it might make more sense to look at a dedicated library like https://github.com/JuliaStrings/utf8proc

**Comment by poire-z** [comment]
The "newlib" stuff looks like it could easily be cut and pasted into our crengine functions with a few search and replace, like you did.

The "musl" stuff looks small but a bit different, less readable. Probably faster, but we don't really mind performance for CSS text-transform.

The "utf8proc" stuff looks a bit more autonomous, and would be a bit of an alternate/duplicate of crengine own (incomplete and wrong in some cases) unicode categorizing. Which initially bothered me.

But I saw some days ago this:
<kbd>![image](https://user-images.githubusercontent.com/24273478/56692968-a6967b80-66e3-11e9-8c57-a99749e01ed1.png)</kbd>
It looks nice, but I never intended for that :)
Thought it was probably some rendering mess I introduced with my LSB/RSB stuff, but no. This char is actually considered to be a _CJKLeftPunctuation_, and handled as such by the CJK code we added, no matter how "hanging punctuation" is set.
So, if "utf8proc" is able to categorize more correctly such CJK punctuation, we could get rid of these kind of things:
https://github.com/koreader/crengine/blob/cf8a533b20cf8be30b30e79aa5ff0a0ed8b7dc61/crengine/src/lvtextfm.cpp#L1392-L1408
and replace them with more accurate categorizing.
Dunno how slower that would be, as it would possibly be done on all chars in a book.

(I would be a bit lost at integrating it in crengine - I guess that's the kind of work @NiLuJe has been doing quite often lately with his fbink :)

> up to date with Unicode 10 instead of 5.2

How did you notice that? Just looking at the codes ?! :) I can't find out much about that...

**Comment by Frenzie** [comment]
> So, if "utf8proc" is able to categorize more correctly such CJK punctuation, we could get rid of these kind of things:

Like these?
https://github.com/JuliaStrings/utf8proc/blob/416749803be5f9d8903c9d95c92153370f890db6/utf8proc.h#L293-L299

But yeah, that's the basic idea behind bringing it up. Just using it for uppercase/lowercase would be an obvious immediate application, but it might open up some other possibilities like that.

> How did you notice that?

I barely glanced at the codes except to find some relevant Polish ones (i.e., the chars around 0x105) which happened to be pretty much right at the top. It says "Based on and tested against Unicode 5.2" in a comment, while musl says Unicode 10 in the changelog.

utf8proc says Unicode 12 in version 2.3 that just came out a couple of weeks ago (2019-03-30).

**Comment by poire-z** [comment]
> Like these?

Well, these don't say anything about them being CJK :(
I guess these could be used to generalized the category that crengine has for the few first 100s unicodes, like in https://github.com/koreader/crengine/pull/237/files#diff-4c9d4af6bc460fb4ab77f7a834614c8c (which would be more efficient than any library call).

But there will always be something missing from any lib, and we want to have other stuff as a flag for a char: OK, we can generalize "AVOID_WRAP_AFTER" on all `UTF8PROC_CATEGORY_PS = 14, /**< Punctuation, open */`, but we will still have to add our own tests to do that too on `$` or `%`, but not on `&`. Or do stuff on some CJK `UTF8PROC_CATEGORY_PS`, but not on non-CJK `UTF8PROC_CATEGORY_PS`...

**Comment by Frenzie** [comment]
Fixed by #5017.
